xwiki-platform-core/xwiki-platform-rendering/xwiki-platform-rendering-xwiki/src/main/java/org/xwiki/rendering/internal/renderer/XWikiDocumentURITitleGenerator.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.rendering.internal.renderer;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.xwiki.component.annotation.Component;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.model.reference.DocumentReferenceResolver;
+import org.xwiki.rendering.listener.reference.ResourceReference;
+import org.xwiki.localization.ContextualLocalizationManager;
+import org.xwiki.rendering.renderer.reference.link.URITitleGenerator;
+
+/**
+ * Generate link titles for document URIs.
+ *
+ * @version $Id$
+ * @since 15.3RC1
+ */
+@Component(hints = {"doc", "page"})
+@Singleton
+public class XWikiDocumentURITitleGenerator implements URITitleGenerator
+{
+    @Inject
+    private ContextualLocalizationManager contextLocalization;
+
+    /**
+     * Used to extract the document name part in a document reference.
+     */
+    @Inject
+    @Named("current")
+    private DocumentReferenceResolver<String> currentDocumentReferenceResolver;
+
+    @Override
+    public String generateCreateTitle(ResourceReference reference)
+    {
+        DocumentReference documentReference =
+            this.currentDocumentReferenceResolver.resolve(reference.getReference());
```

Comment:
This will only work for `DocumentResourceReference` and for other types, it'll cause a bug. For example `SpaceResourceReference` references won't be resolved correctly.